oauth/lib/redirecturiparser.cpp

```diff
@@ -0,0 +1,51 @@
+/****************************************************************************
+**
+** Copyright (c) 2021 Open Mobile Platform LLC
+**
+****************************************************************************/
+
+#include "redirecturiparser_p.h"
+
+#include <QtCore/QUrl>
+#include <QtCore/QUrlQuery>
+#include <QtCore/QStringList>
+#include <QtCore/QString>
+#include <QtCore/QByteArray>
+#include <QtCore/QPair>
+#include <QtCore/QList>
+
+QVariantMap SailfishOS::OAuth::parseRedirectUri(const QString &redirectUri, QString *errorMessage)
+{
+    QVariantMap data;
+
+    const QUrl url(redirectUri);
+    const QUrlQuery query(url.query());
+    const QStringList fragments = url.fragment(QUrl::FullyEncoded).split(QChar('&'));
+
+    QString errorString;
+
+    QList<QPair<QString, QString> > fragmentPairs;
```

Comment:
QUrlQuery has a method queryItems(), that return parsed QList<QPair<QString, QString>>. You don't need to parse it by yourself. Can't you use it here?

Reply:
That queryItems() method is used at the end of the function, but it only allows parsing query items, and doesn't parse key=value pairs from fragments.  This custom parseRedirectUri() method exists to ensure that whether the server provides the result as query items or as encoded key/value pairs in a fragment, we handle both in identical way.

Reply:
Got it. Didn't take into account these uri fragments.